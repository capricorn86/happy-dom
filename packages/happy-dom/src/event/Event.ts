import IEventInit from './IEventInit.js';
import INode from '../nodes/node/INode.js';
import IWindow from '../window/IWindow.js';
import IShadowRoot from '../nodes/shadow-root/IShadowRoot.js';
import IEventTarget from './IEventTarget.js';
import NodeTypeEnum from '../nodes/node/NodeTypeEnum.js';
import { performance } from 'perf_hooks';
import EventPhaseEnum from './EventPhaseEnum.js';
import IDocument from '../nodes/document/IDocument.js';

/**
 * Event.
 */
export default class Event {
	public composed = false;
	public bubbles = false;
	public cancelable = false;
	public defaultPrevented = false;
	public eventPhase: EventPhaseEnum = EventPhaseEnum.none;
	public _immediatePropagationStopped = false;
	public _propagationStopped = false;
	public _target: IEventTarget = null;
	public _currentTarget: IEventTarget = null;
	public timeStamp: number = performance.now();
	public type: string = null;
	public _isInPassiveEventListener = false;
	public NONE = EventPhaseEnum.none;
	public CAPTURING_PHASE = EventPhaseEnum.capturing;
	public AT_TARGET = EventPhaseEnum.atTarget;
	public BUBBLING_PHASE = EventPhaseEnum.bubbling;

	/**
	 * Constructor.
	 *
	 * @param type Event type.
	 * @param [eventInit] Event init.
	 */
	constructor(type: string, eventInit: IEventInit = null) {
		this.type = type;

		if (eventInit) {
			this.bubbles = eventInit.bubbles || false;
			this.cancelable = eventInit.cancelable || false;
			this.composed = eventInit.composed || false;
		}
	}

	/**
	 * Returns target.
	 *
	 * @returns Target.
	 */
	public get target(): IEventTarget {
		return this._target;
	}

	/**
	 * Returns target.
	 *
	 * @returns Target.
	 */
	public get currentTarget(): IEventTarget {
		return this._currentTarget;
	}

	/**
	 * Returns "true" if propagation has been stopped.
	 *
	 * @returns "true" if propagation has been stopped.
	 */
	public get cancelBubble(): boolean {
		return this._propagationStopped;
	}

	/**
	 * Returns composed path.
	 *
	 * @returns Composed path.
	 */
	public composedPath(): IEventTarget[] {
		if (!this._target) {
			return [];
		}

		const composedPath = [];
		let eventTarget: INode | IShadowRoot | IWindow = <INode | IShadowRoot>(<unknown>this._target);

		while (eventTarget) {
			composedPath.push(eventTarget);

			if ((<INode>(<unknown>eventTarget)).parentNode) {
				eventTarget = (<INode>(<unknown>eventTarget)).parentNode;
			} else if (
				this.composed &&
				(<INode>eventTarget).nodeType === NodeTypeEnum.documentFragmentNode &&
				(<IShadowRoot>eventTarget).host
			) {
				eventTarget = (<IShadowRoot>eventTarget).host;
			} else if ((<INode>eventTarget).nodeType === NodeTypeEnum.documentNode) {
				eventTarget = (<IDocument>(<unknown>eventTarget)).defaultView;
			} else {
				break;
			}
		}

		return composedPath;
	}

	/**
	 * Init event.
	 *
	 * @deprecated
	 * @param type Type.
	 * @param [bubbles=false] "true" if it bubbles.
	 * @param [cancelable=false] "true" if it cancelable.
	 */
	public initEvent(type: string, bubbles = false, cancelable = false): void {
		this.type = type;
		this.bubbles = bubbles;
		this.cancelable = cancelable;
	}

	/**
	 * Prevents default.
	 */
	public preventDefault(): void {
		if (!this._isInPassiveEventListener) {
			this.defaultPrevented = true;
		}
	}

	/**
	 * Stops immediate propagation.
	 */
	public stopImmediatePropagation(): void {
		this._immediatePropagationStopped = true;
	}

	/**
	 * Stops propagation.
	 */
	public stopPropagation(): void {
		this._propagationStopped = true;
	}
}
