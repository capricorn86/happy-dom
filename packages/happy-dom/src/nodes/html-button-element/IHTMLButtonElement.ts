import ValidityState from '../../validity-state/ValidityState.js';
import IHTMLElement from '../html-element/IHTMLElement.js';
import IHTMLFormElement from '../html-form-element/IHTMLFormElement.js';
import IHTMLLabelElement from '../html-label-element/IHTMLLabelElement.js';
import INodeList from '../node/INodeList.js';

/**
 * HTML Button Element.
 *
 * Reference:
 * https://developer.mozilla.org/en-US/docs/Web/API/HTMLButtonElement.
 */
export default interface IHTMLButtonElement extends IHTMLElement {
	name: string;
	value: string;
	disabled: boolean;
	type: string;
	formNoValidate: boolean;
	readonly validity: ValidityState;
	readonly form: IHTMLFormElement;
	readonly validationMessage: string;
	readonly labels: INodeList<IHTMLLabelElement>;

	/**
	 * Checks validity.
	 *
	 * @returns Validity.
	 */
	checkValidity(): boolean;

	/**
	 * Reports validity.
	 *
	 * @returns Validity.
	 */
	reportValidity(): boolean;

	/**
	 * Sets validation message.
	 *
	 * @param message Message.
	 */
	setCustomValidity(message: string): void;
}
