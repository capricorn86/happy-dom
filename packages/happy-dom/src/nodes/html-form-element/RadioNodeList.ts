import IElement from '../element/IElement';
import NodeList from '../node/NodeList';
import IRadioNodeList from './IRadioNodeList';

/**
 * RadioNodeList
 *
 * @see https://developer.mozilla.org/en-US/docs/Web/API/RadioNodeList
 */
export default class RadioNodeList extends NodeList<IElement> implements IRadioNodeList {
	/**
	 * Returns value.
	 *
	 * @returns Value.
	 */
	public get value(): string {
		for (const node of this) {
			if (node.checked) {
				return node.value;
			}
		}
		return null;
	}
}
