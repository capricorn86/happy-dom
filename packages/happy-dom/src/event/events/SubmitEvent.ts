import Event from '../Event';
import ISubmitEventInit from './ISubmitEventInit';
import IHTMLElement from '../../nodes/html-element/IHTMLElement';

/**
 * An event triggered by form submit buttons.
 */
export default class SubmitEvent extends Event {
	public readonly submitter: IHTMLElement;

	/**
	 * Constructor.
	 *
	 * @param type Event type.
	 * @param [eventInit] Event init.
	 */
	constructor(type: string, eventInit: ISubmitEventInit = null) {
		super(type, eventInit);

		this.submitter = eventInit?.submitter || null;
	}
}
