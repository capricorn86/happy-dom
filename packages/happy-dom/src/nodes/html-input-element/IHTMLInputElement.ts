import IHTMLElement from '../html-element/IHTMLElement.js';
import IHTMLFormElement from '../html-form-element/IHTMLFormElement.js';
import HTMLInputElementSelectionModeEnum from './HTMLInputElementSelectionModeEnum.js';
import ValidityState from '../../validity-state/ValidityState.js';
import Event from '../../event/Event.js';
import File from '../../file/File.js';
import IFileList from './IFileList.js';
import INodeList from '../node/INodeList.js';
import IHTMLLabelElement from '../html-label-element/IHTMLLabelElement.js';

/**
 * HTML Input Element.
 *
 * Reference:
 * https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement.
 */
export default interface IHTMLInputElement extends IHTMLElement {
	readonly form: IHTMLFormElement;
	readonly labels: INodeList<IHTMLLabelElement>;
	readonly validity: ValidityState;
	formAction: string;
	formMethod: string;
	formNoValidate: boolean;
	defaultChecked: boolean;
	files: IFileList<File>;
	defaultValue: string;
	height: number;
	width: number;
	size: number;
	minLength: number;
	maxLength: number;
	type: string;
	name: string;
	alt: string;
	min: string;
	max: string;
	pattern: string;
	placeholder: string;
	step: string;
	inputmode: string;
	accept: string;
	allowdirs: string;
	autocomplete: string;
	src: string;
	readOnly: boolean;
	disabled: boolean;
	autofocus: boolean;
	required: boolean;
	indeterminate: boolean;
	multiple: boolean;
	checked: boolean;
	value: string;
	selectionStart: number;
	selectionEnd: number;
	selectionDirection: string;
	willValidate: boolean;
	valueAsDate: Date | null;
	valueAsNumber: number;
	validationMessage: string;

	// Events
	oninput: (event: Event) => void | null;
	oninvalid: (event: Event) => void | null;
	onselectionchange: (event: Event) => void | null;

	/**
	 * Sets validation message.
	 *
	 * @param message Message.
	 */
	setCustomValidity(message: string): void;

	/**
	 * Reports validity by dispatching an "invalid" event.
	 */
	reportValidity(): void;

	/**
	 * Selects the text.
	 */
	select(): void;

	/**
	 * Set selection range.
	 *
	 * @param start Start.
	 * @param end End.
	 * @param [direction="none"] Direction.
	 */
	setSelectionRange(start: number, end: number, direction: string): void;

	/**
	 * Set range text.
	 *
	 * @param replacement Replacement.
	 * @param [start] Start.
	 * @param [end] End.
	 * @param [direction] Direction.
	 * @param selectionMode
	 */
	setRangeText(
		replacement: string,
		start?: number,
		end?: number,
		selectionMode?: HTMLInputElementSelectionModeEnum
	): void;

	/**
	 * Checks validity.
	 *
	 * @returns "true" if the field is valid.
	 */
	checkValidity(): boolean;

	/**
	 * Steps up.
	 *
	 * @param [increment] Increment.
	 */
	stepUp(increment?: number): void;

	/**
	 * Steps up.
	 *
	 * @param [increment] Increment.
	 */
	stepDown(increment?: number): void;

	/**
	 * Clones a node.
	 *
	 * @override
	 * @param [deep=false] "true" to clone deep.
	 * @returns Cloned node.
	 */
	cloneNode(deep?: boolean): IHTMLInputElement;
}
