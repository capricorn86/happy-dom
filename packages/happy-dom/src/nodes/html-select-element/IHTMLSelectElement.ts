import IHTMLElement from '../html-element/IHTMLElement';
import IHTMLFormElement from '../html-form-element/IHTMLFormElement';
import IHTMLLabelElement from '../html-label-element/IHTMLLabelElement';
import INodeList from '../node/INodeList';
import IHTMLOptionsCollection from './IHTMLOptionsCollection';
import ValidityState from '../../validity-state/ValidityState';
import Event from '../../event/Event';
import IHTMLOptionElement from '../html-option-element/IHTMLOptionElement';
import IHTMLOptGroupElement from '../html-opt-group-element/IHTMLOptGroupElement';

/**
 * HTML Select Element.
 *
 * Reference:
 * https://developer.mozilla.org/en-US/docs/Web/API/HTMLSelectElement.
 */
export default interface IHTMLSelectElement extends IHTMLElement {
	readonly form: IHTMLFormElement;
	readonly labels: INodeList<IHTMLLabelElement>;
	readonly options: IHTMLOptionsCollection;
	readonly type: string;
	readonly validity: ValidityState;
	readonly willValidate: boolean;
	readonly validationMessage: string;
	autofocus: boolean;
	disabled: boolean;
	length: number;
	selectedIndex: number;
	value: string;
	name: string;
	multiple: boolean;
	required: boolean;

	// Events
	onchange: (event: Event) => void | null;
	oninput: (event: Event) => void | null;

	/**
	 * Adds new option to collection.
	 *
	 * @param element HTMLOptionElement or HTMLOptGroupElement to add.
	 * @param before HTMLOptionElement or index number.
	 */
	add(
		element: IHTMLOptionElement | IHTMLOptGroupElement,
		before?: number | IHTMLOptionElement | IHTMLOptGroupElement
	): void;

	/**
	 * Returns option element by index.
	 *
	 * @param index Index.
	 */
	item(index: number): IHTMLOptionElement | IHTMLOptGroupElement;

	/**
	 * Removes option element from the collection.
	 *
	 * @param index Index.
	 */
	remove(index?: number): void;

	/**
	 * Sets validation message.
	 *
	 * @param message Message.
	 */
	setCustomValidity(message: string): void;

	/**
	 * Checks validity.
	 *
	 * @returns "true" if the field is valid.
	 */
	checkValidity(): boolean;

	/**
	 * Reports validity.
	 *
	 * @returns "true" if the field is valid.
	 */
	reportValidity(): boolean;
}
