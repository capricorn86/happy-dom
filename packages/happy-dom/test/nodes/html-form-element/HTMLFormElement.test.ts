import Window from '../../../src/window/Window.js';
import Document from '../../../src/nodes/document/Document.js';
import HTMLFormElement from '../../../src/nodes/html-form-element/HTMLFormElement.js';
import RadioNodeList from '../../../src/nodes/html-form-element/RadioNodeList.js';
import IHTMLInputElement from '../../../src/nodes/html-input-element/IHTMLInputElement.js';
import Event from '../../../src/event/Event.js';
import SubmitEvent from '../../../src/event/events/SubmitEvent.js';
import IHTMLSelectElement from '../../../src/nodes/html-select-element/IHTMLSelectElement.js';
import IHTMLTextAreaElement from '../../../src/nodes/html-text-area-element/IHTMLTextAreaElement.js';
import IHTMLButtonElement from '../../../src/nodes/html-button-element/IHTMLButtonElement.js';
import { beforeEach, describe, it, expect } from 'vitest';

describe('HTMLFormElement', () => {
	let window: Window;
	let document: Document;
	let element: HTMLFormElement;

	beforeEach(() => {
		window = new Window();
		document = window.document;
		element = <HTMLFormElement>document.createElement('form');
	});

	describe('Object.prototype.toString', () => {
		it('Returns `[object HTMLFormElement]`', () => {
			expect(Object.prototype.toString.call(element)).toBe('[object HTMLFormElement]');
		});
	});

	for (const property of [
		'name',
		'target',
		'action',
		'encoding',
		'enctype',
		'acceptCharset',
		'autocomplete'
	]) {
		describe(`get ${property}()`, () => {
			it('Returns attribute value.', () => {
				expect(element[property]).toBe('');
				element.setAttribute(property, 'value');
				expect(element[property]).toBe('value');
			});
		});

		describe(`set ${property}()`, () => {
			it('Sets attribute value.', () => {
				element[property] = 'value';
				expect(element.getAttribute(property)).toBe('value');
			});
		});
	}

	describe('get noValidate()', () => {
		it('Returns "true" if defined.', () => {
			expect(element.noValidate).toBe(false);
			element.setAttribute('novalidate', '');
			expect(element.noValidate).toBe(true);
		});
	});

	describe('set noValidate()', () => {
		it('Sets attribute value.', () => {
			element.noValidate = true;
			expect(element.getAttribute('novalidate')).toBe('');
		});
	});

	describe('get method()', () => {
		it('Returns attribute value.', () => {
			expect(element.method).toBe('get');
			element.setAttribute('method', 'post');
			expect(element.method).toBe('post');
		});
	});

	describe('set method()', () => {
		it('Sets attribute value.', () => {
			element.method = 'post';
			expect(element.getAttribute('method')).toBe('post');
		});
	});

	describe('get elements()', () => {
		it('Returns control elements.', () => {
			element.innerHTML = `
                <div>
                    <input type="text" name="text1" value="value1">
					<button name="button1" value="value1"></button>
                    <input type="checkbox" name="checkbox1" value="value1">
                    <input type="checkbox" name="checkbox1" value="value2" checked>
                    <input type="checkbox" name="checkbox1" value="value3">
                    <input type="radio" name="radio1" value="value1">
                    <input type="radio" name="radio1" value="value2" checked>
                    <input type="radio" name="radio1" value="value3">
                    <input type="hidden" name="1" value="value1">
                </div>
            `;
			const elements = element.elements;
			const root = element.children[0];

			expect(element.length).toBe(9);
			expect(elements.length).toBe(9);

			expect(element[0] === root.children[0]).toBe(true);
			expect(element[1] === root.children[1]).toBe(true);
			expect(element[2] === root.children[2]).toBe(true);
			expect(element[3] === root.children[3]).toBe(true);
			expect(element[4] === root.children[4]).toBe(true);
			expect(element[5] === root.children[5]).toBe(true);
			expect(element[6] === root.children[6]).toBe(true);
			expect(element[7] === root.children[7]).toBe(true);
			expect(element[8] === root.children[8]).toBe(true);

			expect(elements[0] === root.children[0]).toBe(true);
			expect(elements[1] === root.children[1]).toBe(true);
			expect(elements[2] === root.children[2]).toBe(true);
			expect(elements[3] === root.children[3]).toBe(true);
			expect(elements[4] === root.children[4]).toBe(true);
			expect(elements[5] === root.children[5]).toBe(true);
			expect(elements[6] === root.children[6]).toBe(true);
			expect(elements[7] === root.children[7]).toBe(true);
			expect(elements[8] === root.children[8]).toBe(true);

			expect(elements.item(0) === root.children[0]).toBe(true);
			expect(elements.item(1) === root.children[1]).toBe(true);
			expect(elements.item(2) === root.children[2]).toBe(true);
			expect(elements.item(3) === root.children[3]).toBe(true);
			expect(elements.item(4) === root.children[4]).toBe(true);
			expect(elements.item(5) === root.children[5]).toBe(true);
			expect(elements.item(6) === root.children[6]).toBe(true);
			expect(elements.item(7) === root.children[7]).toBe(true);
			expect(elements.item(8) === root.children[8]).toBe(true);

			const radioNodeList1 = new RadioNodeList();
			const radioNodeList2 = new RadioNodeList();

			radioNodeList1.push(root.children[2]);
			radioNodeList1.push(root.children[3]);
			radioNodeList1.push(root.children[4]);

			radioNodeList2.push(root.children[5]);
			radioNodeList2.push(root.children[6]);
			radioNodeList2.push(root.children[7]);

			expect(element['text1'] === root.children[0]).toBe(true);
			expect(element['button1'] === root.children[1]).toBe(true);
			expect(element['checkbox1']).toEqual(radioNodeList1);
			expect(element['radio1']).toEqual(radioNodeList2);

			expect(elements['text1'] === root.children[0]).toBe(true);
			expect(elements['button1'] === root.children[1]).toBe(true);
			expect(elements['checkbox1']).toEqual(radioNodeList1);
			expect(elements['radio1']).toEqual(radioNodeList2);

			expect(elements.namedItem('text1') === root.children[0]).toBe(true);
			expect(elements.namedItem('button1') === root.children[1]).toBe(true);
			expect(elements.namedItem('checkbox1')).toEqual(radioNodeList1);
			expect(elements.namedItem('checkbox1')?.value).toBe('value2');
			expect(elements.namedItem('radio1')).toEqual(radioNodeList2);
			expect(elements.namedItem('radio1')?.value).toBe('value2');
			expect(elements.namedItem('1')?.value).toBe('value1');

			(<IHTMLInputElement>elements.namedItem('text1')).name = 'text2';
			(<IHTMLInputElement>elements.namedItem('text2')).id = 'text3';

			expect(element['text2'] === root.children[0]).toBe(true);
			expect(element['text3'] === root.children[0]).toBe(true);

			expect(elements['text2'] === root.children[0]).toBe(true);
			expect(elements['text3'] === root.children[0]).toBe(true);

			const anotherElement = document.createElement('div');
			anotherElement.innerHTML = `
                <span>
                    <input type="text" name="anotherText1" value="value1">
                    <button name="anotherButton1" value="value1"></button>
                    <input type="checkbox" name="anotherCheckbox1" value="value1">
                    <input type="checkbox" name="anotherCheckbox1" value="value2" checked>
                    <input type="checkbox" name="anotherCheckbox1" value="value3">
                    <input type="radio" name="anotherRadio1" value="value1">
                    <input type="radio" name="anotherRadio1" value="value2" checked>
                    <input type="radio" name="anotherRadio1" value="value3">
                </span>
            `;

			const anotherRoot = anotherElement.children[0];

			root.appendChild(anotherElement);

			expect(element.length).toBe(17);
			expect(elements.length).toBe(17);

			expect(element[9] === anotherRoot.children[0]).toBe(true);
			expect(element[10] === anotherRoot.children[1]).toBe(true);
			expect(element[11] === anotherRoot.children[2]).toBe(true);
			expect(element[12] === anotherRoot.children[3]).toBe(true);
			expect(element[13] === anotherRoot.children[4]).toBe(true);
			expect(element[14] === anotherRoot.children[5]).toBe(true);
			expect(element[15] === anotherRoot.children[6]).toBe(true);
			expect(element[16] === anotherRoot.children[7]).toBe(true);

			expect(elements[9] === anotherRoot.children[0]).toBe(true);
			expect(elements[10] === anotherRoot.children[1]).toBe(true);
			expect(elements[11] === anotherRoot.children[2]).toBe(true);
			expect(elements[12] === anotherRoot.children[3]).toBe(true);
			expect(elements[13] === anotherRoot.children[4]).toBe(true);
			expect(elements[14] === anotherRoot.children[5]).toBe(true);
			expect(elements[15] === anotherRoot.children[6]).toBe(true);
			expect(elements[16] === anotherRoot.children[7]).toBe(true);

			const anotherRadioNodeList1 = new RadioNodeList();
			const anotherRadioNodeList2 = new RadioNodeList();

			anotherRadioNodeList1.push(anotherRoot.children[2]);
			anotherRadioNodeList1.push(anotherRoot.children[3]);
			anotherRadioNodeList1.push(anotherRoot.children[4]);

			anotherRadioNodeList2.push(anotherRoot.children[5]);
			anotherRadioNodeList2.push(anotherRoot.children[6]);
			anotherRadioNodeList2.push(anotherRoot.children[7]);

			expect(element['anotherText1'] === anotherRoot.children[0]).toBe(true);
			expect(element['anotherButton1'] === anotherRoot.children[1]).toBe(true);
			expect(element['anotherCheckbox1']).toEqual(anotherRadioNodeList1);
			expect(element['anotherRadio1']).toEqual(anotherRadioNodeList2);

			expect(elements['anotherText1'] === anotherRoot.children[0]).toBe(true);
			expect(elements['anotherButton1'] === anotherRoot.children[1]).toBe(true);
			expect(elements['anotherCheckbox1']).toEqual(anotherRadioNodeList1);
			expect(elements['anotherRadio1']).toEqual(anotherRadioNodeList2);

			for (const child of root.children.slice()) {
				if (child !== anotherElement) {
					root.removeChild(child);
				}
			}

			expect(element.length).toBe(8);
			expect(elements.length).toBe(8);

			expect(element[0] === anotherRoot.children[0]).toBe(true);
			expect(element[1] === anotherRoot.children[1]).toBe(true);
			expect(element[2] === anotherRoot.children[2]).toBe(true);
			expect(element[3] === anotherRoot.children[3]).toBe(true);
			expect(element[4] === anotherRoot.children[4]).toBe(true);
			expect(element[5] === anotherRoot.children[5]).toBe(true);
			expect(element[6] === anotherRoot.children[6]).toBe(true);
			expect(element[7] === anotherRoot.children[7]).toBe(true);

			expect(elements[0] === anotherRoot.children[0]).toBe(true);
			expect(elements[1] === anotherRoot.children[1]).toBe(true);
			expect(elements[2] === anotherRoot.children[2]).toBe(true);
			expect(elements[3] === anotherRoot.children[3]).toBe(true);
			expect(elements[4] === anotherRoot.children[4]).toBe(true);
			expect(elements[5] === anotherRoot.children[5]).toBe(true);
			expect(elements[6] === anotherRoot.children[6]).toBe(true);
			expect(elements[7] === anotherRoot.children[7]).toBe(true);

			expect(element['text1'] === undefined).toBe(true);
			expect(element['button1'] === undefined).toBe(true);
			expect(element['checkbox1'] === undefined).toBe(true);
			expect(element['radio1'] === undefined).toBe(true);

			expect(elements['text1'] === undefined).toBe(true);
			expect(elements['button1'] === undefined).toBe(true);
			expect(elements['checkbox1'] === undefined).toBe(true);
			expect(elements['radio1'] === undefined).toBe(true);
		});

		it('Returns control elements using the same name as properties and methods of the HTMLCollection class.', () => {
			element.innerHTML = `
                <div>
                    <input type="text" name="length" value="value1">
                    <input type="checkbox" name="namedItem" value="value1">
                    <input type="checkbox" name="namedItem" value="value2" checked>
                    <input type="checkbox" name="namedItem" value="value3">
                    <input type="hidden" name="push" value="value1">
                </div>
            `;
			const elements = element.elements;
			const root = element.children[0];

			expect(element.length).toBe(5);
			expect(elements.length).toBe(5);

			expect(element[0] === root.children[0]).toBe(true);
			expect(element[1] === root.children[1]).toBe(true);
			expect(element[2] === root.children[2]).toBe(true);
			expect(element[3] === root.children[3]).toBe(true);
			expect(element[4] === root.children[4]).toBe(true);

			expect(elements[0] === root.children[0]).toBe(true);
			expect(elements[1] === root.children[1]).toBe(true);
			expect(elements[2] === root.children[2]).toBe(true);
			expect(elements[3] === root.children[3]).toBe(true);
			expect(elements[4] === root.children[4]).toBe(true);

			expect(elements.item(0) === root.children[0]).toBe(true);
			expect(elements.item(1) === root.children[1]).toBe(true);
			expect(elements.item(2) === root.children[2]).toBe(true);
			expect(elements.item(3) === root.children[3]).toBe(true);
			expect(elements.item(4) === root.children[4]).toBe(true);

			const radioNodeList = new RadioNodeList();
			radioNodeList.push(root.children[1]);
			radioNodeList.push(root.children[2]);
			radioNodeList.push(root.children[3]);

			expect(typeof elements.push).toBe('function');
			expect(typeof elements.namedItem).toBe('function');
			expect(elements.namedItem('length') === root.children[0]).toBe(true);
			expect(elements.namedItem('namedItem')).toEqual(radioNodeList);
			expect(elements.namedItem('push') === root.children[4]).toBe(true);

			const children = root.children.slice();

			for (const child of children) {
				root.removeChild(child);
			}

			expect(element.length).toBe(0);

			for (const child of children) {
				root.appendChild(child);
			}

			expect(element.length).toBe(5);
			expect(elements.length).toBe(5);

			expect(elements[0] === root.children[0]).toBe(true);
			expect(elements[1] === root.children[1]).toBe(true);
			expect(elements[2] === root.children[2]).toBe(true);
			expect(elements[3] === root.children[3]).toBe(true);
			expect(elements[4] === root.children[4]).toBe(true);

			expect(typeof elements.push).toBe('function');
			expect(typeof elements.namedItem).toBe('function');
			expect(elements.namedItem('length') === root.children[0]).toBe(true);
			expect(elements.namedItem('namedItem')).toEqual(radioNodeList);
			expect(elements.namedItem('push') === root.children[4]).toBe(true);
		});
	});

	describe('submit()', () => {
		it('Does nothing.', () => {
			element.submit();
		});
	});

	describe('requestSubmit()', () => {
		it('Validates form and triggers a "submit" event when valid.', () => {
			element.innerHTML = `
                <div>
                    <input type="text" name="text1" required>
					<button name="button1"></button>
                    <input type="checkbox" name="checkbox1" value="value1" required>
                    <input type="checkbox" name="checkbox1" value="value2">
                    <input type="checkbox" name="checkbox1" value="value3">
                    <input type="radio" name="radio1" value="value1" required>
                    <input type="radio" name="radio1" value="value2" required>
                    <input type="radio" name="radio1" value="value3" required>
                </div>
            `;

			document.body.appendChild(element);

			const root = element.children[0];
			let submitEvent: Event | null = null;

			element.addEventListener('submit', (event: Event) => (submitEvent = event));

			element.noValidate = true;
			element.requestSubmit();

			expect(submitEvent).toBeInstanceOf(SubmitEvent);
			expect((<SubmitEvent>(<unknown>submitEvent)).type).toBe('submit');
			expect((<SubmitEvent>(<unknown>submitEvent)).submitter).toBe(element);

			submitEvent = null;

			element.noValidate = false;
			element.requestSubmit();

			expect(submitEvent).toBe(null);

			(<IHTMLInputElement>root.children[0]).value = 'value';
			(<IHTMLInputElement>root.children[2]).click();
			(<IHTMLInputElement>root.children[6]).click();

			element.requestSubmit(<IHTMLButtonElement>root.children[1]);

			expect((<SubmitEvent>(<unknown>submitEvent)).type).toBe('submit');
			expect((<SubmitEvent>(<unknown>submitEvent)).submitter).toBe(root.children[1]);
		});

		it('Skips validating if a submitter is sent that has "formNoValidate" set to "true".', () => {
			element.innerHTML = `
                <div>
                    <input type="text" name="text1" required>
                    <input type="checkbox" name="checkbox1" value="value1" required>
                    <input type="checkbox" name="checkbox1" value="value2">
                    <input type="checkbox" name="checkbox1" value="value3">
                    <input type="radio" name="radio1" value="value1" required>
                    <input type="radio" name="radio1" value="value2" required>
                    <input type="radio" name="radio1" value="value3" required>
					<input type="submit" name="button1"></input>
                </div>
            `;

			document.body.appendChild(element);

			const root = element.children[0];
			const submitter = <IHTMLInputElement>root.children[7];
			let submitEvent: Event | null = null;

			element.addEventListener('submit', (event: Event) => (submitEvent = event));

			submitter.formNoValidate = true;
			element.requestSubmit(submitter);

			expect((<SubmitEvent>(<unknown>submitEvent)).type).toBe('submit');
		});
	});

	describe('reset()', () => {
		it('Resets the form.', () => {
			element.innerHTML = `
                <div>
                    <input type="text" name="text1" value="Default value">
                    <select>
                        <option value="value1"></option>
                        <option value="value2" selected></option>
                        <option value="value3"></option>
                    </select>
                    <textarea name="textarea1">Default value</textarea>
                    <input type="checkbox" name="checkbox1" value="value1">
                    <input type="checkbox" name="checkbox1" value="value2" checked>
                    <input type="checkbox" name="checkbox1" value="value3">
                    <input type="radio" name="radio1" value="value1">
                    <input type="radio" name="radio1" value="value2" checked>
                    <input type="radio" name="radio1" value="value3">
                </div>
            `;

			document.body.appendChild(element);

			const root = element.children[0];
			let resetEvent: Event | null = null;

			(<IHTMLInputElement>root.children[0]).value = 'New value';
			(<IHTMLSelectElement>root.children[1]).value = 'value3';
			(<IHTMLTextAreaElement>root.children[2]).value = 'New value';
			(<IHTMLInputElement>root.children[3]).click();
			(<IHTMLInputElement>root.children[5]).click();
			(<IHTMLInputElement>root.children[7]).click();

			element.addEventListener('reset', (event: Event) => (resetEvent = event));

			element.reset();

			expect((<Event>(<unknown>resetEvent)).type).toBe('reset');

			expect((<IHTMLInputElement>root.children[0]).value).toBe('Default value');
			expect((<IHTMLSelectElement>root.children[1]).value).toBe('value2');
			expect((<IHTMLTextAreaElement>root.children[2]).value).toBe('Default value');

			expect((<IHTMLInputElement>root.children[3]).checked).toBe(false);
			expect((<IHTMLInputElement>root.children[4]).checked).toBe(true);
			expect((<IHTMLInputElement>root.children[5]).checked).toBe(false);

			expect((<IHTMLInputElement>root.children[6]).checked).toBe(false);
			expect((<IHTMLInputElement>root.children[7]).checked).toBe(true);
			expect((<IHTMLInputElement>root.children[8]).checked).toBe(false);
		});
	});

	for (const method of ['checkValidity', 'reportValidity']) {
		describe(`${method}()`, () => {
			it('Validates the form.', () => {
				element.innerHTML = `
                <div>
                    <input type="text" name="text1" required>
                    <select required>
                        <option></option>
                    </select>
                    <textarea name="textarea1" required></textarea>
                    <input type="checkbox" name="checkbox1" value="value1" required>
                    <input type="checkbox" name="checkbox1" value="value2" required>
                    <input type="checkbox" name="checkbox1" value="value3" required>
                    <input type="radio" name="radio1" value="value1" required>
                    <input type="radio" name="radio1" value="value2" required>
                    <input type="radio" name="radio1" value="value3" required>
                </div>
            `;

				document.body.appendChild(element);

				const root = element.children[0];
				let invalidEvents: Event[] = [];

				element.addEventListener('invalid', (event: Event) => invalidEvents.push(event));

				expect(element[method]()).toBe(false);
				expect(invalidEvents.length).toBe(7);
				invalidEvents = [];

				(<IHTMLInputElement>root.children[0]).value = 'value';
				(<IHTMLSelectElement>root.children[1]).options[0].value = 'value';
				(<IHTMLTextAreaElement>root.children[2]).value = 'value';
				(<IHTMLInputElement>root.children[3]).click();
				(<IHTMLInputElement>root.children[4]).click();
				(<IHTMLInputElement>root.children[5]).click();
				(<IHTMLInputElement>root.children[7]).click();

				expect(element[method]()).toBe(true);
				expect(invalidEvents.length).toBe(0);
			});
		});
	}
});
