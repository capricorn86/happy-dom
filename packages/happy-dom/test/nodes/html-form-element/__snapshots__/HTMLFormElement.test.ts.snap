// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HTMLFormElement > constructor() > Matches snapshot. 1`] = `
<form>
  
                
  <input
    name="text1"
    type="text"
    value="value1"
  />
  
                
  <input
    name="text2"
    type="hidden"
    value="value2"
  />
  
                
  <input
    checked=""
    name="checkbox1"
    type="checkbox"
    value="value1"
  />
  
                
  <input
    name="checkbox2"
    type="checkbox"
    value="value2"
  />
  
                
  <input
    name="radio1"
    type="radio"
    value="value1"
  />
  
                
  <input
    checked=""
    name="radio1"
    type="radio"
    value="value2"
  />
  
                
  <input
    name="radio1"
    type="radio"
    value="value3"
  />
  
                
  <input
    name="button1"
    type="submit"
  />
  
            
</form>
`;
